### PR TITLE
Creating/randomnizing only 1 of γ or δ is enough

### PR DESCRIPTION
**Pull request by ytrezq**
While randomizing both or works, taking only 1 of them is enough. In fact this what Zcash’s Sapling’s trusted setup do : after phase 1, both values are equal to the G₂ generators point leading to the attack described in this article. But after phase2, only values linked to δ are randomized, thus leaving public inputs to their phase1 value with the Generator point in G₂.

[See the relevant Zcash’s trusted setup source’s code](https://github.com/ebfull/phase2/blob/master/src/lib.rs#L808)

### CODE REVIEW

**Comment by jeffreyscholz** [comment]
Indeed, gamma2 is G₂ here: https://github.com/tornadocash/tornado-core/blob/1ef6a263ac6a0e476d063fcb269a9df65a1bd56a/contracts/Verifier.sol#L176C25-L176C113

Will give this a review soon, thank you!

File: content/groth16/en/groth16.md [hunk 1]
```diff
@@ -299,10 +299,10 @@ The equation above is valid, but the witness does not necessarily satisfy the or
...
-To avoid the problem above, the trusted setup introduces new scalars $\gamma$ and $\delta$ to force $\Psi_{\ell+1}$ to $\Psi_m$ to be separate from $\Psi_1$ to $\Psi_{\ell}$. To do this, the trusted setup divides (multiplies by the modular inverse) the private terms (that constitute $[C]_1$) by $\delta$ and the public terms (that constitute $[X]_1$, the sum the verifier computes) by $\gamma$.
+### Introducing $\gamma$ and/or $\delta$
+To avoid the problem above, the trusted setup introduces a new scalar : $\gamma$ or $\delta$ to force $\Psi_{\ell+1}$ to $\Psi_m$ to be separate from $\Psi_1$ to $\Psi_{\ell}$. To do this, the trusted setup divides (multiplies by the modular inverse) the private terms (that constitute $[C]_1$) by $\delta$ and/or the public terms (that constitute $[X]_1$, the sum the verifier computes) by $\gamma$.
 
-Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$.
+Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$. Having $\delta$ and $\gamma$ having an unknown discrete logarithm relation is enough to avoid any forgeries.
```
Comment (RareSkills) [inline content/groth16/en/groth16.md]: Can we reword this to "If either $\delta$ or $\gamma$ have an unknown discrete log, then the forgery described earlier is avoided. In practice, $\gamma$ is simply $G_2$ and $\delta$ is randomly generated during the trusted setup. In the Tornado Cash verifier, the [`gamma2` elliptic curve point](https://github.com/tornadocash/tornado-core/blob/1ef6a263ac6a0e476d063fcb269a9df65a1bd56a/contracts/Verifier.sol#L176C25-L176C113) is the $\mathbb{G}_2$ generator. Similarly, in the [ZCash trusted setup](https://github.com/ebfull/phase2/blob/master/src/lib.rs#L808) (back when ZCash used Groth16), $\gamma$ was simply $G_2$ and $\delta$ was randomly chosen.

**Review by RareSkills — CHANGES_REQUESTED** [review]
Hey this is good, just want to incorporate more context.

File: content/groth16/en/groth16.md (hunk 1 above)
Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills I created this pull request with the power to let you commit what I did…
Zcash is still using Groth16 : Zcash on the technical level works as 3 cryptocurrencies : the bitcoin 1, the sapling 1 ; the orchad 1 : each of them as power of emmission in the others. Moreover, Tornado Cash is irrelevant since they are using the same code Zcash for their trusted setup ceremony (just adapting the existing code to the different curve).

Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills : some Tornado cash keys (like classic versions deployed on other chains and nova) that don’t rely on the Zcash’s trusted setup code do have both δ and γ randomly generated as a result. Most groth16 implementation do update both δ and γ.

File: content/groth16/en/groth16.md [hunk 2]
```diff
@@ -299,10 +299,10 @@ The equation above is valid, but the witness does not necessarily satisfy the or
...
+### Introducing $\gamma$ and/or $\delta$
+To avoid the problem above, the trusted setup introduces a new scalar : $\gamma$ or $\delta$ to force $\Psi_{\ell+1}$ to $\Psi_m$ to be separate from $\Psi_1$ to $\Psi_{\ell}$. To do this, the trusted setup divides (multiplies by the modular inverse) the private terms (that constitute $[C]_1$) by $\delta$ and/or the public terms (that constitute $[X]_1$, the sum the verifier computes) by $\gamma$.
 
-Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$.
+Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$. If either $\delta$ and $\gamma$ have an unknown discrete logarithm, then the forgery described earlier along possible other methods are avoided. This method is typically used in Zcash’s Sapling’s‑based trusted setups where $\gamma$ is simply $G_2$ and $\delta$ is still randomly generated.
```
Comment (RareSkills) [inline content/groth16/en/groth16.md]: "typically" isn't the right word because it was either used or it wasn't. I think "This method was used in Zcash's Sapling's-based [trusted setups](https://github.com/ebfull/phase2/blob/master/src/lib.rs#L808)"

we need to include the hyperlink because we can't expect the reader to look it up themselves.

**Review by RareSkills — CHANGES_REQUESTED** [review]
see the latest suggestion

File: content/groth16/en/groth16.md (hunk 2 above)
Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills : done.

File: content/groth16/en/groth16.md [hunk 3]
```diff
@@ -299,10 +299,10 @@ The equation above is valid, but the witness does not necessarily satisfy the or
...
+### Introducing $\gamma$ and/or $\delta$
+To avoid the problem above, the trusted setup introduces a new scalar : $\gamma$ or $\delta$ to force $\Psi_{\ell+1}$ to $\Psi_m$ to be separate from $\Psi_1$ to $\Psi_{\ell}$. To do this, the trusted setup divides (multiplies by the modular inverse) the private terms (that constitute $[C]_1$) by $\delta$ and/or the public terms (that constitute $[X]_1$, the sum the verifier computes) by $\gamma$.
 
-Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$.
+Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$. If either $\delta$ and $\gamma$ have an unknown discrete logarithm, then the forgery described earlier along possible other methods are avoided. This method was used in Zcash’s Sapling’s‑based [trusted setups](https://github.com/ebfull/phase2/blob/master/src/lib.rs#L808) where $\gamma$ is simply left to $G_2$ after phase1 and $\delta$ is still randomly generated in phase2.
```
Comment (RareSkills) [inline content/groth16/en/groth16.md]: Unfortunately, we'll have to leave off "in phase2"

The reader does not yet have the context of "phase1" and "phase2" in the context of a trusted setup, and this would raise questions.

Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills Ok. Change done. I think it’s important to make notice that G₂ is the first value used to build the circuit.

File: content/groth16/en/groth16.md [hunk 4]
```diff
@@ -299,10 +299,10 @@ The equation above is valid, but the witness does not necessarily satisfy the or
...
+### Introducing $\gamma$ and/or $\delta$
+To avoid the problem above, the trusted setup introduces a new scalar : $\gamma$ or $\delta$ to force $\Psi_{\ell+1}$ to $\Psi_m$ to be separate from $\Psi_1$ to $\Psi_{\ell}$. To do this, the trusted setup divides (multiplies by the modular inverse) the private terms (that constitute $[C]_1$) by $\delta$ and/or the public terms (that constitute $[X]_1$, the sum the verifier computes) by $\gamma$.
 
-Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$.
+Since the $h(\tau)t(\tau)$ term is embedded in $[C]_1$, those terms also need to be divided by $\delta$. If either $\delta$ and $\gamma$ have an unknown discrete logarithm, then the forgery described earlier along possible other methods are avoided. This method was used in Zcash’s Sapling’s‑based [trusted setups](https://github.com/ebfull/phase2/blob/master/src/lib.rs#L808) where $\gamma$ is simply left to $G_2$ and $\delta$ is still randomly generated from $G_2$.
```
Comment (RareSkills) [inline content/groth16/en/groth16.md]: the phrase "randomly generated from $G_2$" doesn't quite describe a concrete operation.

I believe you might have meant "randomly generate from $G_2$" means "we multiplied $G_2$ by a random number" but it isn't quite saying that.

Some alternatives:
- is randomly chosen during the trusted setup, then published as $[\gamma G_2]$
- is some other random point in the $\mathbb{G}_2$ elliptic curve group

**Review by RareSkills — CHANGES_REQUESTED** [review]

File: content/groth16/en/groth16.md (hunk 4 above)
Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills No. What I’m meaning is values are generated according to G₂ then such linked values are updated from G₂ when G₂ become randomized into γ or δ. For example, the points used in G₁ for public inputs are generated first according to G₂ and then when γ is randomized, the public inputs points are updated.

Do you have a better way to say this ?

Comment (RareSkills) [reply content/groth16/en/groth16.md]: I think where I'm confused is "generated according to G₂"

What does it mean to generate according to a certain point (the generator in this case)?

Comment (ytrezq) [reply content/groth16/en/groth16.md]: @RareSkills you generate values first in G₁ in phase1, then after phase1 when you apply randomness on δ or γ in phase2,  you apply the same scalar on the previously generated G₁ values in order to keep proportions.

How to explain that without talking about the 2 phases of ᴍᴘᴄ ?

**Review by RareSkills — APPROVED** [review]

**Comment by ytrezq** [comment]
@RareSkills when those changes will posted on the medium blog?
